opentelemetry-sdk/src/logs/growable_array.rs

```diff
@@ -0,0 +1,323 @@
+    ///
+    /// The iterator yields elements from the internal array (`initial`) first, followed by elements
+    /// from the vector (`additional`) if present. This allows for efficient iteration over both
+    /// stack-allocated and heap-allocated portions.
+    ///
+    #[inline]
+    pub(crate) fn iter(&self) -> GrowableArrayIter<'_, T, MAX_STACK_CAPACITY> {
+        if self.additional.is_none() || self.additional.as_ref().unwrap().is_empty() {
+            GrowableArrayIter::StackOnly {
+                iter: self.initial.iter().take(self.count),
+            }
+        } else {
+            GrowableArrayIter::Mixed {
+                stack_iter: self.initial.iter().take(self.count),
+                heap_iter: self.additional.as_ref().unwrap().iter(),
+            }
+        }
+    }
+}
+
+// Implement `IntoIterator` for `GrowableArray`
+impl<T: Default + Clone + PartialEq, const INITIAL_CAPACITY: usize> IntoIterator
+    for GrowableArray<T, INITIAL_CAPACITY>
+{
+    type Item = T;
+    type IntoIter = GrowableArrayIntoIter<T, INITIAL_CAPACITY>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        if self.additional.is_none() || self.additional.as_ref().unwrap().is_empty() {
+            GrowableArrayIntoIter::StackOnly {
+                iter: self.initial.into_iter().take(self.count),
+            }
+        } else {
+            GrowableArrayIntoIter::Mixed {
+                stack_iter: self.initial.into_iter().take(self.count),
+                heap_iter: self.additional.unwrap().into_iter(),
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+/// Iterator for consuming a `GrowableArray`.
+///
+/// The iterator consumes the `GrowableArray`, yielding each element by value.
+/// It first iterates over the elements in the internal array (`initial`), and then,
+/// if present, over the elements in the vector (`additional`).
+///
+pub(crate) enum GrowableArrayIntoIter<T: Default + Clone + PartialEq, const INITIAL_CAPACITY: usize>
+{
+    /// stackonly
+    StackOnly {
+        /// iter
+        iter: std::iter::Take<IntoIter<T, INITIAL_CAPACITY>>,
+    },
+    /// hybrid
+    Mixed {
+        /// stack_iter
+        stack_iter: std::iter::Take<IntoIter<T, INITIAL_CAPACITY>>,
+        /// heap_iter
+        heap_iter: std::vec::IntoIter<T>,
+    },
+}
+
+impl<T: Default + Clone + PartialEq, const INITIAL_CAPACITY: usize> Iterator
+    for GrowableArrayIntoIter<T, INITIAL_CAPACITY>
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            GrowableArrayIntoIter::StackOnly { iter } => iter.next(),
+            GrowableArrayIntoIter::Mixed {
+                stack_iter,
+                heap_iter,
+            } => stack_iter.next().or_else(|| heap_iter.next()),
+        }
+    }
+}
+
+// Implement `IntoIterator` for a reference to `GrowableArray`
+impl<'a, T: Default + Clone + PartialEq + 'a, const INITIAL_CAPACITY: usize> IntoIterator
```

Comment:
Could we remove this? Anyone trying to iterate over references of the attributes could simply use `.iter()`, right?

```diff
    #[test]
    fn test_ref_iter() {
        let mut collection = GrowableArray::<i32>::new();
        for i in 0..15 {
            collection.push(i);
        }
-        let iter = &collection;
+        let iter = collection.iter();
        let mut count = 0;
        for value in iter {
            assert_eq!(*value, count);
            count += 1;
        }
        assert_eq!(count, 15);
    }
```

Reply:
Yes makes sense. We can always bring this back later if needed.